lnrpc/signrpc/signer_server.go

```diff
@@ -506,38 +488,76 @@ func (s *Server) VerifyMessage(ctx context.Context,
 
 // DeriveSharedKey returns a shared secret key by performing Diffie-Hellman key
 // derivation between the ephemeral public key in the request and the node's
-// key specified in the key_loc parameter (or the node's identity private key
-// if no key locator is specified):
-//     P_shared = privKeyNode * ephemeralPubkey
+// key specified in the key_desc parameter. Then key_family of the key locator
+// must be supplied, and either a key_index of the key locator or a raw public
+// key is expected in the key_desc.
 // The resulting shared public key is serialized in the compressed format and
 // hashed with sha256, resulting in the final key length of 256bit.
 func (s *Server) DeriveSharedKey(_ context.Context, in *SharedKeyRequest) (
 	*SharedKeyResponse, error) {
 
-	if len(in.EphemeralPubkey) != 33 {
-		return nil, fmt.Errorf("ephemeral pubkey must be " +
-			"serialized in compressed format")
-	}
-	ephemeralPubkey, err := btcec.ParsePubKey(
-		in.EphemeralPubkey, btcec.S256(),
-	)
+	// Check that EphemeralPubkey is valid.
+	ephemeralPubkey, err := parseRawKeyBytes(in.EphemeralPubkey)
 	if err != nil {
-		return nil, fmt.Errorf("unable to parse pubkey: %v", err)
+		return nil, fmt.Errorf("error in ephemeral pubkey: %v", err)
+	}
+
+	// Check for backward compatibility. The caller either specifies the old
+	// key_loc field, or the new key_desc field, but not both.
+	if in.KeyDesc != nil && in.KeyLoc != nil {
+		return nil, fmt.Errorf("use either key_desc or key_loc")
+	}
+
+	// We extract two params, rawKeyBytes and keyLoc. Notice their initial
+	// values will be overwritten if not using the deprecated RPC param.
+	var rawKeyBytes []byte
+	keyLoc := in.KeyLoc
+	if in.KeyDesc != nil {
+		keyLoc = in.KeyDesc.GetKeyLoc()
+		rawKeyBytes = in.KeyDesc.GetRawKeyBytes()
 	}
 
-	// By default, use the node identity private key.
-	locator := keychain.KeyLocator{
-		Family: keychain.KeyFamilyNodeKey,
-		Index:  0,
+	// Check that the key family is supplied. Note that since the empty value of
+	// the key family is 0, and the KeyFamilyMultiSig is also 0, we cannot
+	// validate whether the key family is supplied or not. Instead we check to
+	// make sure the key locator is supplied, and the key family is valid.
+	if keyLoc == nil {
+		return nil, fmt.Errorf("missing param key_loc")
```

Comment:
This should instead set the `keyLoc` variable to the node identity family to be consistent with the description of the service:
```
		keyLoc = &KeyLocator{
			KeyFamily: int32(keychain.KeyFamilyNodeKey),
			KeyIndex:  0,
		}
```

Reply:
Changed. Does this mean the `key_family` is not a mandatory field?

Reply:
It is mandatory _if_ either of the key locators is not `nil` (the root `key_loc` or the `key_loc` inside of `key_desc`).
But we can't really check if it's set, as 0 is a valid key.